### PR TITLE
Add `format` field to EGL surface backing store

### PR DESCRIPTION
**Pull request by vially**
Trying to use the `GL_BGRA8_EXT` format for the EGL surface backing store on *desktop* OpenGL platforms would fail at: https://github.com/flutter/engine/blob/bde314b856850155d0678632b62c23c84f49afab/shell/platform/embedder/embedder.cc#L869

This seems to be a known issue and both the Linux and Windows embedders have some logic to pick a different format depending on the OpenGL context information:

https://github.com/flutter/engine/blob/088dcfa6cf934088413c5d8af754de1ef7b0d1ca/shell/platform/windows/compositor_opengl.cc#L23-L34

https://github.com/flutter/engine/blob/088dcfa6cf934088413c5d8af754de1ef7b0d1ca/shell/platform/linux/fl_framebuffer.cc#L81-L104

This pull-request gets rid of the hard-coded `GL_BGRA8_EXT` format and makes it configurable by adding a `format` field to the `FlutterOpenGLSurface` struct.

_Disclaimer_: This has only been tested on desktop Linux (Wayland) using the `GR_GL_RGBA8` format which seemed to work as expected.

This change is related to the recently introduced EGL surface backing store: https://github.com/flutter/flutter/issues/58363

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by loic-sharma** [comment]
@vially Thanks for opening this patch! I think it would be better if we followed the other backing stores' pattern and instead introduced a `format` parameter to the config to allow the embedder to choose its format. See this discussion: https://github.com/flutter/engine/pull/43683/files#r1263989268

FYI, we just recently landed https://github.com/flutter/engine/pull/54329 which moves away from `kN32_SkColorType` and instead uses `FlutterFormatToSkColorType` to convert the config's format to an `SkColorType`. We will likely want to use that here as well.

cc @chinmaygarde

**Comment by vially** [comment]
@loic-sharma Thanks for the review. I have updated the implementation with your suggestion by adding a `format` field to the `FlutterOpenGLSurface` embedder struct (similar to the other backing stores).

Does it make sense to also add some logic to set the default format to `GL_BGRA8_EXT` (when set to `0`/unset), just to maintain compatibility with the previous logic?

Or maybe backwards compatibility is not a real concern in this case, since the EGL surface backing store was only merged a few days ago?

/cc @ardera

**Comment by ardera** [comment]
> @loic-sharma Thanks for the review. I have updated the implementation with your suggestion by adding a `format` field to the `FlutterOpenGLSurface` embedder struct (similar to the other backing stores).
> 
> Does it make sense to also add some logic to set the default format to `GL_BGRA8_EXT` (when set to `0`/unset), just to maintain compatibility with the previous logic?
> 
> Or maybe backwards compatibility is not a real concern in this case, since the EGL surface backing store (#43683) was only merged a few days ago?
> 
> /cc @ardera

Oof thanks for the fix, I wasn't aware of this.

IMO I think it's also a good idea to document this limitation embedder API, e.g.:
```C
typedef struct {
  // . . .

  /// The surface format (example GL_RGBA8).
  ///
  /// Due to limitations in Skia, this needs to be GL_BGRA8_EXT when either:
  ///   - GL_EXT_texture_format_BGRA8888 is present or
  ///   - GL_APPLE_texture_format_BGRA8888 is present and GL version is larger than 3.0
  /// otherwise it should be GL_RGBA8.
  uint32_t format;
} FlutterOpenGLSurface;
```
Maybe also with a TODO link for an issue. It's a bit ugly to write that there, but it's even more annoying to run into the assert, and having to dig through engine / other embedder sources to find the fix.

I'd say backwards compatibility is not a concern yet, it hasn't even reached stable yet. We also ignored backwards compatibility for https://github.com/flutter/engine/pull/38847 (and that was when the software pixel formats were already in stable) (though that was API not ABI)

**Comment by loic-sharma** [comment]
> I'd say backwards compatibility is not a concern yet, it hasn't even reached stable yet. 

Agreed. I'd say go ahead and make the `format` parameter required. @chinmaygarde has the final word here though :)

> IMO I think it's also a good idea to document this limitation in the embedder API, e.g.:

Agreed documenting allowed formats this is a good idea, but I wouldn't prescribe the logic to choose the format as that's likely highly dependent on the embedder and its platform. Perhaps a comment like:


```cpp
typedef struct {
  ...

  /// The surface format.
  ///
  /// Allowed values:
  ///   - GL_RGBA8
  ///   - GL_BGRA8_EXT
  uint32_t format;
} FlutterOpenGLSurface;
```

**Comment by vially** [comment]
Thank you both. I have updated the field comment with your suggestions to list the allowed values. I've also updated the test producer class to set to format in order to make the tests pass.

File: shell/platform/embedder/tests/embedder_test_backingstore_producer.cc
Comment (loic-sharma) [inline shell/platform/embedder/tests/embedder_test_backingstore_producer.cc:262]: Should this be `GL_BGRA8_EXT`? That would better match the previous test behavior. Or does that cause tests to fail on your machine?

**Comment by loic-sharma** [comment]
Are there any tests we could add for this?

File: shell/platform/embedder/tests/embedder_test_backingstore_producer.cc
Comment (vially) [reply shell/platform/embedder/tests/embedder_test_backingstore_producer.cc]: I have updated it to `GL_RGBA8_EXT` and the tests still pass on my machine. That being said, I wasn't sure what's the most suitable GL header (which contains this value) to be included in these tests, so I just copied the value instead.

I'm happy to replace the hard-coded value with a proper import if someone can point me to the correct header file that should be used.

**Comment by vially** [comment]
> Are there any tests we could add for this?

I have added a new test for validating that passing an invalid format will result in an invalid Skia surface. The "happy" path of valid format resulting in valid Skia surface is already covered by the original tests from #43683, so I think the common code-paths are now covered.

It would be nice if the tests covered some of the edge-cases too (e.g.: trying to use `GL_BGRA8_EXT` when it's not available) but I haven't found an easy way to do that.

**Comment by jtmcdole** [comment]
@loic-sharma - friendly ping

**Comment by iamsergio** [comment]
I've tested this patch on Linux/X11 and now GL finally works! Before only GLES worked.

**Comment by chinmaygarde** [comment]
@loic-sharma Can you tell us how to proceed here?

**Comment by loic-sharma** [comment]
Apologies for the delay, this fell off my radar. I'll take another look tomorrow.

File: shell/platform/embedder/embedder.cc
Comment (loic-sharma) [inline shell/platform/embedder/embedder.cc:868]: I believe we shipped a Flutter release that had `FlutterOpenGLSurface` without a `format` member.

Could we wrap this in a `SAFE_ACCESS` that defaults to `GL_BGRA8_EXT` if `FlutterOpenGLSurface` doesn't have a `format` member?

**Review by loic-sharma — APPROVED** [review]
These changes look good to me (modulo [this minor comment](https://github.com/flutter/engine/pull/54499#discussion_r1774305410)).

I'm not an expert in this area, please get an approval from @chinmaygarde before merging this.

File: shell/platform/embedder/embedder.cc
Comment (vially) [reply shell/platform/embedder/embedder.cc]: This should now be fixed in latest revision.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by loic-sharma — APPROVED** [review]
Re-LGTM with updates - thanks for the wonderful contribution!

**Pull request merged**
